File: aoc2025-go/days/day06/day06.go
package day06

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"aoc2025-go/common"
	"aoc2025-go/parse"
)

const IS_TEST = false

type Problem struct {
	nums []int
	op   string
}

type Grid struct {
	rows [][]int
	ops  []string
}

func parseRow(line string) []int {
	fields := strings.Fields(line)
	nums := make([]int, len(fields))
	for i, f := range fields {
		nums[i], _ = strconv.Atoi(f)
	}
	return nums
}

func parseGrid(input []string) Grid {
	numRows := len(input) - 1
	rows := make([][]int, numRows)
	for i := range numRows {
		rows[i] = parseRow(input[i])
	}
	return Grid{rows: rows, ops: strings.Fields(input[numRows])}
}

func parseProblems(input []string) []Problem {
	var problems []Problem
	var currentCols []string

	width := len(input[0])
	for i := range width {
		var b strings.Builder
		for j := range input {
			if i < len(input[j]) {
				b.WriteByte(input[j][i])
			}
		}
		col := b.String()

		allSpace := strings.Count(col, " ") == len(col)
		if !allSpace {
			currentCols = append(currentCols, col)
		}
		if (allSpace || i == width-1) && len(currentCols) > 0 {
			problems = append(problems, buildProblem(currentCols))
			currentCols = nil
		}
	}
	return problems
}

func buildProblem(cols []string) Problem {
	first := cols[0]
	op := string(first[len(first)-1])
	cols[0] = first[:len(first)-1]

	nums := make([]int, 0, len(cols))
	for _, col := range cols {
		trimmed := strings.TrimSpace(col)
		if trimmed == "" {
			continue
		}
		num, err := strconv.Atoi(trimmed)
		if err != nil {
			panic(err)
		}
		nums = append(nums, num)
	}
	return Problem{nums: nums, op: op}
}

func applyOp(op string, nums []int) int {
	if op == "+" {
		sum := 0
		for _, n := range nums {
			sum += n
		}
		return sum
	}
	product := 1
	for _, n := range nums {
		product *= n
	}
	return product
}

func Part1(input []string) string {
	grid := parseGrid(input)
	sum := 0
	for i, op := range grid.ops {
		col := make([]int, len(grid.rows))
		for j, row := range grid.rows {
			col[j] = row[i]
		}
		sum += applyOp(op, col)
	}
	return fmt.Sprintf("%d", sum)
}

func Part2(input []string) string {
	problems := parseProblems(input)
	sum := 0
	for _, p := range problems {
		sum += applyOp(p.op, p.nums)
	}
	return fmt.Sprintf("%d", sum)
}

func Run() {
	input := parse.Lines(6, IS_TEST)

	t1s := time.Now()
	p1 := Part1(input)
	t1 := time.Since(t1s)

	t2s := time.Now()
	p2 := Part2(input)
	t2 := time.Since(t2s)

	common.RunDay(6, p1, p2, t1, t2)
}

File: aoc2025-zig/src/days/day06.zig
const std = @import("std");
const common = @import("../common.zig");
const parse = @import("../parse.zig");

const Problem = struct {
    nums: []i64,
    op: u8,
};

fn parseRow(line: []const u8, allocator: std.mem.Allocator) ![]i64 {
    var nums: std.ArrayList(i64) = .empty;
    var it = std.mem.tokenizeScalar(u8, line, ' ');
    while (it.next()) |token| {
        const n = try std.fmt.parseInt(i64, token, 10);
        try nums.append(allocator, n);
    }
    return nums.toOwnedSlice(allocator);
}

fn applyOp(op: u8, nums: []const i64) i64 {
    if (op == '+') {
        var sum: i64 = 0;
        for (nums) |n| sum += n;
        return sum;
    }
    var product: i64 = 1;
    for (nums) |n| product *= n;
    return product;
}

pub fn part1(input: [][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    // Last row is ops; all prior rows are number rows.
    const num_rows = input.len - 1;

    const rows = try allocator.alloc([]i64, num_rows);
    for (input[0..num_rows], 0..) |line, i| {
        rows[i] = try parseRow(line, allocator);
    }

    var ops: std.ArrayList(u8) = .empty;
    var ops_it = std.mem.tokenizeScalar(u8, input[num_rows], ' ');
    while (ops_it.next()) |token| {
        try ops.append(allocator, token[0]);
    }

    const col_buf = try allocator.alloc(i64, num_rows);
    var sum: i64 = 0;
    for (ops.items, 0..) |op, i| {
        for (rows, 0..) |row, j| col_buf[j] = row[i];
        sum += applyOp(op, col_buf);
    }

    return std.fmt.allocPrint(allocator, "{d}", .{sum});
}

pub fn part2(input: [][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    const width = input[0].len;
    var problems: std.ArrayList(Problem) = .empty;
    var current_cols: std.ArrayList([]u8) = .empty;

    for (0..width) |i| {
        // Build vertical slice: one char per row at column i.
        const col = try allocator.alloc(u8, input.len);
        for (input, 0..) |row, j| {
            col[j] = if (i < row.len) row[i] else ' ';
        }

        var all_space = true;
        for (col) |c| {
            if (c != ' ') {
                all_space = false;
                break;
            }
        }

        if (!all_space) try current_cols.append(allocator, col);

        if ((all_space or i == width - 1) and current_cols.items.len > 0) {
            // First col's last char is the operator; strip it before parsing.
            const first = current_cols.items[0];
            const op = first[first.len - 1];
            first[first.len - 1] = ' ';

            var nums: std.ArrayList(i64) = .empty;
            for (current_cols.items) |col_str| {
                const trimmed = std.mem.trim(u8, col_str, " ");
                if (trimmed.len == 0) continue;
                const n = try std.fmt.parseInt(i64, trimmed, 10);
                try nums.append(allocator, n);
            }

            try problems.append(allocator, Problem{
                .nums = try nums.toOwnedSlice(allocator),
                .op = op,
            });
            current_cols = .empty;
        }
    }

    var sum: i64 = 0;
    for (problems.items) |p| sum += applyOp(p.op, p.nums);

    return std.fmt.allocPrint(allocator, "{d}", .{sum});
}

pub fn run(io: std.Io, allocator: std.mem.Allocator) !void {
    const input = try parse.lines(io, allocator, 6, false);

    const t1s = std.Io.Timestamp.now(io, .awake);
    const p1 = try part1(input, allocator);
    const t1e = std.Io.Timestamp.now(io, .awake);

    const t2s = std.Io.Timestamp.now(io, .awake);
    const p2 = try part2(input, allocator);
    const t2e = std.Io.Timestamp.now(io, .awake);

    common.runDay(io, 6, p1, p2, t1s, t1e, t2s, t2e);
}
